### PR TITLE
feat: Expose renderer spellcheck API

### PR DESCRIPTION
**Pull request by lishid**
#### Description of Change

Exposes renderer side browser spellcheck API to the javascript environment.
Resolves https://github.com/electron/electron/issues/22829

Rationale: Electron-based apps that uses CodeMirror (a hugely popular editor) are unable to benefit from the built-in spellchecker since CodeMirror does not use a ContentEditable (everything is faked). CodeMirror exposes a way to highlight errors, but given that we are currently unable to ask for a spellcheck from javascript, all Electron+CodeMirror apps use a js-based spellcheck library like https://github.com/cfinke/Typo.js which is considerably slower than the built-in native spellchecker.

This PR adds the following API:

- `isWordMisspelled(word: string): boolean`
- `getWordSuggestions(word: string): string[]`

**Current WIP adds the methods to the global `process` object which is just so I can test that it works.**
I've tested this and it works. Now to figure out where to put the API.

Ideally I would like this to be a module on `require('electron')`, similar to [IPCRenderer](https://github.com/electron/electron/blob/master/shell/renderer/api/electron_api_ipc_renderer.cc).
In practice, I have not found a way to pass the `WeakPtr<Spellcheck>` to [the renderer API stack](https://github.com/electron/electron/blob/master/lib/renderer/api/module-list.ts). I was also unable to find a reference back to `ElectronRendererClient`/`RenderClientBase` from an API module.
If someone with experience can give me some tips on how to achieve it that would be great!

Also currently missing tests and documentation.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).
- [x] This is **NOT A BREAKING CHANGE**. Breaking changes may not be merged to master until 11-x-y is branched.

#### Release Notes

Notes: Added spellcheck API to renderer.

### CODE REVIEW

**Comment by codebytere** [comment]
cc @MarshallOfSound

File: shell/renderer/electron_renderer_client.cc
Comment (MarshallOfSound) [inline shell/renderer/electron_renderer_client.cc:196]: Hm, so this definitely shouldn't be on `process`.  My guess is it should probably be on something like `webFrame` 🤔

Comment (lishid) [reply shell/renderer/electron_renderer_client.cc:196]: Yeah agreed. I'm debating whether to put it on `webFrame` or adding its own module, something like `require('electron').spellcheck`.

Here's what I wrote in the PR description:

> In practice, I have not found a way to pass the `WeakPtr<Spellcheck>` to [the renderer API stack](https://github.com/electron/electron/blob/master/lib/renderer/api/module-list.ts). I was also unable to find a reference back to `ElectronRendererClient`/`RenderClientBase` from an API module.
> If someone with experience can give me some tips on how to achieve it that would be great!

**Comment by lishid** [comment]
@MarshallOfSound Quick ping on this. Can you please advise on how I can get a reference from:
[RendererClientBase](https://github.com/electron/electron/blob/master/shell/renderer/renderer_client_base.h#L84) to [the renderer API modules](https://github.com/electron/electron/blob/master/lib/renderer/api/module-list.ts) such as [IPCRenderer](https://github.com/electron/electron/blob/master/shell/renderer/api/electron_api_ipc_renderer.cc)?

This is blocking further progress on this PR. Thanks.

File: shell/renderer/electron_renderer_client.cc
Comment (zcbenz) [reply shell/renderer/electron_renderer_client.cc:196]: @lishid You can find some examples on adding new modules under `shell/renderer/api/` folder. 

Adding the APIs to `webFrame` would be simplest and you only need to move the code to `shell/renderer/api/electron_api_web_frame.cc`.

Comment (lishid) [reply shell/renderer/electron_renderer_client.cc:196]: @zcbenz Thank you for the help! In my working copy I was already able to move the code & API to `webFrame`, however, I was unable to find a reference to [the spellchecker object](https://github.com/electron/electron/blob/master/shell/renderer/renderer_client_base.h#L84) which is located on RefererClientBase. Any tips? It seems that all the samples I've seen only references `content::RenderFrame` gotten from a [v8::Context](https://github.com/electron/electron/blob/master/shell/renderer/api/electron_api_web_frame.cc#L93)

Comment (zcbenz) [reply shell/renderer/electron_renderer_client.cc:196]: Since the `RendererClientBase` is guaranteed to have only one instance per process, I think you can add a `RendererClientBase::Get()` method to return the global instance, like what we do in `ElectronBrowserMainParts::Get`.

Chromium also does a similar thing for `ChromeExtensionsRendererClient`, so this pattern should be fine.

Comment (lishid) [reply shell/renderer/electron_renderer_client.cc:196]: Great having some code I can reference makes this a lot easier. Will update the PR soon, thanks again!

**Comment by lishid** [comment]
Did a quick test and it seems to work pretty well.
I'll start working on tests and documentation if this looks all good 🎉

**Comment by lishid** [comment]
Also I can't seem to be able to view the lint results from CircleCI without "logging in with Github" and giving it read/write access to all my repos. The details link gives a 404 "You may have been logged out."

Curious if this is intentionally configured, and if there's an alternate way to check the result details.

File: shell/renderer/api/electron_api_web_frame.cc
Comment (zcbenz) [inline shell/renderer/api/electron_api_web_frame.cc:678]: Can you move this to anonymous namespace? i.e. where the `GetRenderFrame` are defined.

**Comment by zcbenz** [comment]
The lint error is:

```patch
$ node ./script/lint.js && npm run lint:clang-format && npm run lint:docs
--- a/shell/renderer/api/electron_api_web_frame.cc
+++ b/shell/renderer/api/electron_api_web_frame.cc
@@ -675,9 +675,9 @@
   return stats;
 }
 
-bool SpellCheckWord(v8::Isolate* isolate, 
+bool SpellCheckWord(v8::Isolate* isolate,
                     v8::Local<v8::Value> window,
-                    const std::string& word, 
+                    const std::string& word,
                     std::vector<base::string16>* optional_suggestions) {
   size_t start;
   size_t length;
```

It seems that you need to remove some trailing whitespaces.

I'm not sure why the lint CI result is not visible to you. /cc @jkleinsc

**Comment by zcbenz** [comment]
Current implementation looks good to me, and the API will be discussed at next wg-api's meeting.

Can you also add some documentation for the APIs?

**Comment by lishid** [comment]
Sounds good, will do.

**Comment by jkleinsc** [comment]
@lishid @zcbenz I checked with CircleCI and in regards to viewing CircleCi results (eg lint results), you will need to login to CircleCI to see the details.  You can see the overall status on GitHub, but viewing the details requires logging into CircleCI.  You should be able to pick and choose what repos you give CircleCI access to.

**Comment by lishid** [comment]
@jkleinsc I see. It seems like the only choice given was "All repos" or "Public repos only"

image

Choosing "Public repos only" asks for read-write access to all public repos:

image

IIRC the problem is with Github's OAuth not having granular control over specific repos, but I guess in the meantime it shouldn't be much trouble registering a separate account if access control is an issue.

**Review by nornagon — CHANGES_REQUESTED** [review]
Please add a test!

**Comment by lishid** [comment]
Test and docs have both been added. Excited to finally get native spellchecker support for our CodeMirror editor!
Let me know if this PR still needs any tweaking.

**Comment by jkleinsc** [comment]
The API WG reviewed this on Sept 21, 2020.

**Comment by lishid** [comment]
Quick follow-up: is there anything still blocking this PR?

**Review by nornagon — APPROVED** [review]

File: shell/renderer/api/electron_api_web_frame.cc
Comment (nornagon) [inline shell/renderer/api/electron_api_web_frame.cc:122]: Hang on, I think we can just specify `kNoTag` for the tag? Which would mean we wouldn't need to fetch the render frame, which makes me think maybe this API shouldn't be on `webFrame` at all?

**Review by nornagon — CHANGES_REQUESTED** [review]

File: shell/renderer/api/electron_api_web_frame.cc
Comment (nornagon) [reply shell/renderer/api/electron_api_web_frame.cc:122]: hm... https://source.chromium.org/chromium/chromium/src/+/master:components/spellcheck/renderer/spellcheck_provider.cc;l=310;drc=c83ae1d48c762639e626be9be0c77a209fdea359 is a reference that uses the routing id. there are other places in the code that use `kNoTag`. cc @MarshallOfSound for further review

Comment (nornagon) [reply shell/renderer/api/electron_api_web_frame.cc:122]: It seems to be used in the macOS platform spellchecker impl: https://source.chromium.org/chromium/chromium/src/+/master:components/spellcheck/browser/spellcheck_platform_mac.mm;l=201;drc=da08b7363d0947b0e7d98c5282b4cca513f62100

**Review by nornagon — APPROVED** [review]
switching my review back to approved because i'm satisfied that the routing id is probably the right thing to be passing here.

File: shell/renderer/electron_renderer_client.cc
Comment (miniak) [inline shell/renderer/electron_renderer_client.cc:49]: `self_ = nullptr`

Comment (lishid) [reply shell/renderer/electron_renderer_client.cc:49]: I could add this, but it didn't seem necessary given it wasn't added for ElectronBrowserMainParts https://github.com/electron/electron/blob/01dc50d8295690058fbde34b540e3b44b486bdbc/shell/browser/electron_browser_main_parts.cc#L205

**Comment by lishid** [comment]
Quick follow-up: is there anything still blocking this PR? Seems like it's ready to merge.

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Looks like the newly added tests are failing on Linux

**Comment by lishid** [comment]
> Looks like the newly added tests are failing on Linux

Good catch, seems like I forgot to add the `spellcheck` flag to the test runner.

**Comment by jkleinsc** [comment]
The API WG meeting approved this API at our Oct 5, 2020 meeting.

**Comment by lishid** [comment]
Anyone has an idea why the test script doesn't seems to have spellchecker properly enabled?

I've enabled them here:
https://github.com/electron/electron/commit/c160b8e35fb6278e7e921bcd4fdef609725c95e3#diff-391ce3c5c52c03cc68b42bd9c0fdc0deR115

I've also added a delay so the dictionary has time to load:
https://github.com/electron/electron/commit/c160b8e35fb6278e7e921bcd4fdef609725c95e3#diff-9260609f068ed314da5e8498971a6457R31

Neither seems to have loaded the spellchecker dictionary properly for the test suite...

**Comment by lishid** [comment]
@MarshallOfSound Since you [wrote the original tests for spellchecker](https://github.com/electron/electron/commit/4323b6d618d41376c7fcef85752ebff48602651d) would it be possible to take a look at why the tests here aren't properly setup to enable spellchecker? Thanks in advance!
https://github.com/electron/electron/commit/c160b8e35fb6278e7e921bcd4fdef609725c95e3#diff-9260609f068ed314da5e8498971a6457

**Comment by zcbenz** [comment]
@lishid Have you tested this API on Linux?

**Comment by lishid** [comment]
> @lishid Have you tested this API on Linux?

Yes I have tested the API (Ubuntu 18.04) and it does work from the developer console when loading the app via `e start`.

**Comment by lishid** [comment]
I'm on a slightly older nightly but I don't think that should have affected things:
image

**Comment by lishid** [comment]
Just pulled the latest nightly and spent the whole day rebuilding - Seems like it still works correctly from the developer console.

**Comment by zcbenz** [comment]
What if you run the Electron tests instead? Running code from devtools may behave differently from when running tests.

**Comment by lishid** [comment]
I did run the test and as expected it does fail just like it did in CI.

I've made the test window show up and added a `<textarea>` with spellchecker turned on - it doesn't seem to work there either.

**Comment by lishid** [comment]
I tried various functions and commands but the app seems to "think" spellchecker is enabled overall, the renderer's command line has the `--enable-spellcheck` flag, etc.

Running the functions in the console (while the test is running) doesn't seem to work. Weird given it works in `e start`. I'm running the tests using `e tests --runners=remote`.

**Comment by zcbenz** [comment]
I think Chromium probably requires a textarea to present in _current_ page, and a context menu to be opened to actually initialize the spell checker.

So to make the test pass, you can probably dynamically add a textarea to the page and popup a context menu on it first.

**Comment by lishid** [comment]
I don't think that's the case: I've added a `<textarea>` with a misspelled word, and I tried right clicking on it while the tests were running. I'd expect it should turn red with squiggly line but nothing happened. Subsequent calls to the spellchecker API also fails.

Also in my `e start` testing, which is literally just a clone of `electron-quick-start`, it doesn't have a `<textarea>` but the spellchecker API works right away.

**Comment by lishid** [comment]
Would it be ok to move the tests to https://github.com/electron/electron/blob/master/spec-main/spellchecker-spec.ts instead? I can try if that works.

**Comment by zcbenz** [comment]
I'm good with moving the tests to `spellchecker-spec.ts`.

**Comment by lishid** [comment]
Apologies for the delay. New test is in and I have verified that it works on my Ubuntu 18.04 box. Hopefully the CI is green and we should be getting closer to a merge.

**Review by jkleinsc — APPROVED** [review]
LGTM... the mac arm64 CI failures are not related to this PR.  Rebasing this PR to the latest on master will resolve those failures.

**Comment by lishid** [comment]
I did a rebase which seemed to have fixed the specific build issue with mac, but latest master has some other build failures :(

Anything else I need to do before a merge can happen? 🚀

**Comment by zcbenz** [comment]
The CI is failing on master branch, we need to fix it first. https://github.com/electron/electron/pull/25992

**Review by zcbenz — APPROVED** [review]

**Comment by craftzdog** [comment]
Congrats!!!!!

**Comment by lishid** [comment]
@zcbenz Thanks for the merge!
I'm not too familiar with how the Electron team's roadmap planning works, but I'm curious to know which release version of Electron would contain these changes?
It doesn't seem like it's being backported to v11 so I'm guessing we're aiming for v12?

Thanks again for all the help on this PR!

**Comment by zcbenz** [comment]
Yeah it will be included in v12, for backporting to v11 please check https://github.com/electron/governance/blob/master/wg-releases/feature-backport-requests.md.

**Comment by lishid** [comment]
Sounds good.

> By requesting backport of a feature, the submitter of the PR is assuming responsibility for this feature on the requested branches.

I can take responsibility for this PR if that means we can backport it to v11. Given there's still over 3 weeks until v11 goes stable, let me know what I should do to initiate the backport request.

**Comment by zcbenz** [comment]
I have added this PR to @electron/wg-releases 's meeting items, once it is approved you can start backporting it to 11-x.

**Comment by lishid** [comment]
Seems like we got less than a week to go before v11 hits stable so I'm guessing a backport isn't likely to happen, is that correct?

**Comment by jkleinsc** [comment]
@lishid the @electron/wg-releases missed this one, but we will vote on it.  It probably won't make 11.0.0, but if approved it could go in as a 11.1.0 release.

**Comment by daviddev1** [comment]
What's the current state of this request, when is it going to be implemented into Electron?

**Pull request merged**
